File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
          /*int maxLength = 0;
          for(int i = 0;i<s.length();i++){
            HashSet<Character> set = new HashSet<>();//BRUTE FORCE APPROACH
            for(int j = i;j<s.length();j++){          //TIME COMPLEXITY:0(n^2)
                char ch = s.charAt(j);
                if(set.contains(ch)){
                 break;
                }
            
            set.add(ch);
            maxLength = Math.max(maxLength,j - i+1);
          }
        }
        return maxLength;
    }
          
    }*/

        int left = 0;
        int maxLength = 0;
        Set<Character> set = new HashSet<>();
        for(int right = 0;right<s.length();right++){//TIME COMPLEXITY:0(n)
            while(set.contains(s.charAt(right))){
                set.remove(s.charAt(left));
                left++;
            }
                set.add(s.charAt(right));
                maxLength = Math.max(maxLength,right-left+1);//OPTIMIZED APPROACH

        }
        return maxLength;
        }
        
    }

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
'''class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        max_length = 0
        n = len(s)

        for i in range(n):

            for j in range(i, n):

                substring = s[i:j+1]

                if len(substring) == len(set(substring)):
                    max_length = max(max_length, len(substring))//time complexity:

        return max_length'''
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        char_set = set()

        left = 0
        max_length = 0

        for right in range(len(s)):

            while s[right] in char_set:
                char_set.remove(s[left])
                left += 1

            char_set.add(s[right])

            max_length = max(max_length, right - left + 1)

        return max_length